File: Webfuel.App/ClientApp/src/app/api/report-design.api.ts
import { EventEmitter, Injectable, inject } from '@angular/core';
import { Observable, tap } from 'rxjs';
import { ApiService, ApiOptions } from '../core/api.service';
import { ActivatedRouteSnapshot, ResolveFn, RouterStateSnapshot } from '@angular/router';
import { IDataSource } from 'shared/common/data-source';
import { InsertReportColumn, ReportDesign, UpdateReportColumn, DeleteReportColumn, InsertReportFilter, UpdateReportFilter, DeleteReportFilter, ReportSchema, LookupReferenceField, QueryResult, ReportMapEntity, ReportArgument } from './api.types';

@Injectable()
export class ReportDesignApi {
    constructor(private apiService: ApiService) { }
    
    public insertReportColumn (body: InsertReportColumn, options?: ApiOptions): Observable<ReportDesign> {
        return this.apiService.request<InsertReportColumn, ReportDesign>("POST", "api/report-design/insert-column", body, options);
    }
    
    public updateReportColumn (body: UpdateReportColumn, options?: ApiOptions): Observable<ReportDesign> {
        return this.apiService.request<UpdateReportColumn, ReportDesign>("POST", "api/report-design/update-column", body, options);
    }
    
    public deleteReportColumn (body: DeleteReportColumn, options?: ApiOptions): Observable<ReportDesign> {
        return this.apiService.request<DeleteReportColumn, ReportDesign>("POST", "api/report-design/delete-column", body, options);
    }
    
    public insertReportFilter (body: InsertReportFilter, options?: ApiOptions): Observable<ReportDesign> {
        return this.apiService.request<InsertReportFilter, ReportDesign>("POST", "api/report-design/insert-filter", body, options);
    }
    
    public updateReportFilter (body: UpdateReportFilter, options?: ApiOptions): Observable<ReportDesign> {
        return this.apiService.request<UpdateReportFilter, ReportDesign>("POST", "api/report-design/update-filter", body, options);
    }
    
    public deleteReportFilter (body: DeleteReportFilter, options?: ApiOptions): Observable<ReportDesign> {
        return this.apiService.request<DeleteReportFilter, ReportDesign>("POST", "api/report-design/delete-filter", body, options);
    }
    
    public getReportSchema (params: { reportProviderId: string }, options?: ApiOptions): Observable<ReportSchema> {
        return this.apiService.request<undefined, ReportSchema>("GET", "api/report-design/schema/" + params.reportProviderId + "", undefined, options);
    }
    
    public lookupReferenceField (body: LookupReferenceField, options?: ApiOptions): Observable<QueryResult<ReportMapEntity>> {
        return this.apiService.request<LookupReferenceField, QueryResult<ReportMapEntity>>("POST", "api/report-design/schema/lookup-reference-field", body, options);
    }
    
    public generateArguments (body: ReportDesign, options?: ApiOptions): Observable<Array<ReportArgument>> {
        return this.apiService.request<ReportDesign, Array<ReportArgument>>("POST", "api/report-design/generate-arguments", body, options);
    }
}

File: Webfuel.Api/Api/Report/ReportDesignApi.cs
using MediatR;
using Microsoft.AspNetCore.Mvc;
using Webfuel.Domain;
using Webfuel.Reporting;

namespace Webfuel.App
{
    [ApiService]
    public static class ReportDesignApi
    {
        public static void RegisterEndpoints(IEndpointRouteBuilder app)
        {
            // Report Design Modification (note these have no authentication requirements as they don't save any state)

            app.MapPost("api/report-design/insert-column", InsertReportColumn);
            app.MapPost("api/report-design/update-column", UpdateReportColumn);
            app.MapPost("api/report-design/delete-column", DeleteReportColumn);

            app.MapPost("api/report-design/insert-filter", InsertReportFilter);
            app.MapPost("api/report-design/update-filter", UpdateReportFilter);
            app.MapPost("api/report-design/delete-filter", DeleteReportFilter);

            // Helpers

            app.MapGet("api/report-design/schema/{reportProviderId:guid}", GetReportSchema);
            app.MapPost("api/report-design/schema/lookup-reference-field", LookupReferenceField);
            app.MapPost("api/report-design/generate-arguments", GenerateArguments);
        }

        // Report Design Modification

        public static Task<ReportDesign> InsertReportColumn([FromBody] InsertReportColumn command, IMediator mediator)
        {
            return mediator.Send(command);
        }

        public static Task<ReportDesign> UpdateReportColumn([FromBody] UpdateReportColumn command, IMediator mediator)
        {
            return mediator.Send(command);
        }

        public static Task<ReportDesign> DeleteReportColumn([FromBody] DeleteReportColumn command, IMediator mediator)
        {
            return mediator.Send(command);
        }

        public static Task<ReportDesign> InsertReportFilter([FromBody] InsertReportFilter command, IMediator mediator)
        {
            return mediator.Send(command);
        }

        public static Task<ReportDesign> UpdateReportFilter([FromBody] UpdateReportFilter command, IMediator mediator)
        {
            return mediator.Send(command);
        }

        public static Task<ReportDesign> DeleteReportFilter([FromBody] DeleteReportFilter command, IMediator mediator)
        {
            return mediator.Send(command);
        }

        // Helpers

        public static ReportSchema GetReportSchema(Guid reportProviderId, IReportDesignService reportDesignService)
        {
            return reportDesignService.GetReportSchema(reportProviderId);
        }

        public static Task<QueryResult<ReferenceLookup>> LookupReferenceField([FromBody] LookupReferenceField command, IReportDesignService reportDesignService)
        {
            return reportDesignService.QueryReferenceField(command.ReportProviderId, command.FieldId, command.Query);
        }

        public static Task<List<ReportArgument>> GenerateArguments([FromBody] ReportDesign design, IReportDesignService reportDesignService)
        {
            return reportDesignService.GenerateArguments(design);
        }
    }

    public class LookupReferenceField
    {
        public required Guid ReportProviderId { get; set; }

        public required Guid FieldId { get; set; }

        public required Query Query { get; set; }
    }
}
